flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterQueryableStateTest.java

```diff
@@ -0,0 +1,343 @@
+
+    static {
+        JOB_VERTEX_1 = new JobVertex("v1");
+        JOB_VERTEX_1.setParallelism(PARALLELISM);
+        JOB_VERTEX_1.setInvokableClass(AbstractInvokable.class);
+
+        JOB_VERTEX_2 = new JobVertex("v2");
+        JOB_VERTEX_2.setParallelism(PARALLELISM);
+        JOB_VERTEX_2.setInvokableClass(AbstractInvokable.class);
+
+        // explicit configuration required for #testRegisterAndUnregisterKvState
+        JOB_VERTEX_1.setMaxParallelism(16);
+        JOB_VERTEX_2.setMaxParallelism(16);
+
+        final SlotSharingGroup slotSharingGroup = new SlotSharingGroup();
+        JOB_VERTEX_1.setSlotSharingGroup(slotSharingGroup);
+        JOB_VERTEX_2.setSlotSharingGroup(slotSharingGroup);
+
+        JOB_GRAPH = new JobGraph(JOB_VERTEX_1, JOB_VERTEX_2);
+        JOB_GRAPH.setJobType(JobType.STREAMING);
+    }
+
+    @BeforeClass
+    public static void setupClass() {
+        rpcService = new TestingRpcService();
+    }
+
+    @After
+    public void teardown() throws Exception {
+        rpcService.clearGateways();
+    }
+
+    @AfterClass
+    public static void teardownClass() {
+        if (rpcService != null) {
+            rpcService.stopService();
+            rpcService = null;
+        }
+    }
+
+    @Test
+    public void testRequestKvStateWithoutRegistration() throws Exception {
+        final JobMaster jobMaster = new JobMasterBuilder(JOB_GRAPH, rpcService).createJobMaster();
+
+        jobMaster.start();
+
+        final JobMasterGateway jobMasterGateway = jobMaster.getSelfGateway(JobMasterGateway.class);
+
+        registerSlotsRequiredForJobExecution(jobMasterGateway);
+
+        try {
+            // lookup location
+            try {
+                jobMasterGateway.requestKvStateLocation(JOB_GRAPH.getJobID(), "unknown").get();
+                fail("Expected to fail with UnknownKvStateLocation");
+            } catch (Exception e) {
+                assertTrue(
+                        ExceptionUtils.findThrowable(e, UnknownKvStateLocation.class).isPresent());
+            }
+        } finally {
+            RpcUtils.terminateRpcEndpoint(jobMaster, testingTimeout);
+        }
+    }
+
+    @Test
+    public void testRequestKvStateOfWrongJob() throws Exception {
+        final JobMaster jobMaster = new JobMasterBuilder(JOB_GRAPH, rpcService).createJobMaster();
+
+        jobMaster.start();
+
+        final JobMasterGateway jobMasterGateway = jobMaster.getSelfGateway(JobMasterGateway.class);
+
+        registerSlotsRequiredForJobExecution(jobMasterGateway);
+
+        try {
+            // lookup location
+            try {
+                jobMasterGateway.requestKvStateLocation(new JobID(), "unknown").get();
+                fail("Expected to fail with FlinkJobNotFoundException");
+            } catch (Exception e) {
+                assertTrue(
+                        ExceptionUtils.findThrowable(e, FlinkJobNotFoundException.class)
+                                .isPresent());
```

Comment:
Maybe we can add a `FlinkMatchers` matcher to cover the case where we are looking for a specific type. The benefit would be that we see the exception if it does not match. Now we would only see that the condition is false.